x/payment/pfd.go

```diff
@@ -0,0 +1,97 @@
+package payment
+
+import (
+	"context"
+	"google.golang.org/grpc"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdk_tx "github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+
+	"github.com/celestiaorg/celestia-app/x/payment/types"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+var (
+	// shareSizes includes all the possible share sizes of the given data
+	// that the signer must sign over.
+	shareSizes = []uint64{16, 32, 64, 128}
+)
+
+// SubmitPayForData constructs, signs and synchronously submits a PayForData
+// transaction, returning a sdk.TxResponse upon submission.
+func SubmitPayForData(
+	ctx context.Context,
+	signer *types.KeyringSigner,
+	conn *grpc.ClientConn,
+	nID namespace.ID,
+	data []byte,
+	gasLim uint64,
+) (*sdk.TxResponse, error) {
+	pfd, err := BuildPayForData(ctx, signer, conn, nID, data, gasLim)
+	if err != nil {
+		return nil, err
+	}
+
+	signed, err := SignPayForData(signer, pfd, types.SetGasLimit(gasLim))
+	if err != nil {
+		return nil, err
+	}
+
+	rawTx, err := signer.EncodeTx(signed)
+	if err != nil {
+		return nil, err
+	}
+
+	txResp, err := types.BroadcastTx(ctx, conn, sdk_tx.BroadcastMode_BROADCAST_MODE_BLOCK, rawTx)
+	if err != nil {
+		return nil, err
+	}
+	return txResp.TxResponse, nil
+}
```

Comment:
finally, the prophesied good api for submitting a payForData has come :raised_hands: